src/raftstore/coprocessor/split_check/half.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use rocksdb::DB;
+
+use util::config::ReadableSize;
+
+use super::super::{Coprocessor, ObserverContext, SplitCheckObserver};
+use super::Status;
+
+const BUCKET_NUMBER_LIMIT: usize = 1024;
+const BUCKET_SIZE_LIMIT_MB: u64 = 512;
+
+#[derive(Default)]
+pub struct HalfStatus {
+    buckets: Vec<Vec<u8>>,
+    cur_bucket_size: u64,
+    each_bucket_size: u64,
+}
+
+impl HalfStatus {
+    fn new(each_bucket_size: u64) -> HalfStatus {
+        HalfStatus {
+            each_bucket_size: each_bucket_size,
+            ..Default::default()
+        }
+    }
+
+    pub fn push_data(&mut self, key: &[u8], value_size: u64) {
+        let current_len = key.len() as u64 + value_size;
+        if self.buckets.is_empty() || self.cur_bucket_size >= self.each_bucket_size {
+            self.buckets.push(key.to_vec());
+            self.cur_bucket_size = 0;
+        }
+        self.cur_bucket_size += current_len;
+    }
+
+    pub fn split_key(mut self) -> Option<Vec<u8>> {
+        let mid = self.buckets.len() / 2;
+        if mid == 0 {
+            None
+        } else {
+            Some(self.buckets.swap_remove(mid))
+        }
+    }
+}
+
+pub struct HalfCheckObserver {
+    half_split_bucket_size: u64,
+}
+
+impl HalfCheckObserver {
+    pub fn new(region_size_limit: u64) -> HalfCheckObserver {
+        let mut half_split_bucket_size = region_size_limit / BUCKET_NUMBER_LIMIT as u64;
+        let bucket_size_limit = ReadableSize::mb(BUCKET_SIZE_LIMIT_MB).0;
+        if half_split_bucket_size == 0 {
+            half_split_bucket_size = 1;
+        } else if half_split_bucket_size > bucket_size_limit {
+            half_split_bucket_size = bucket_size_limit;
+        }
+        HalfCheckObserver {
+            half_split_bucket_size: half_split_bucket_size,
+        }
+    }
+}
+
+impl Coprocessor for HalfCheckObserver {}
+
+impl SplitCheckObserver for HalfCheckObserver {
+    fn new_split_check_status(
+        &self,
+        _ctx: &mut ObserverContext,
```

Comment:
I think we can `get_region_approximate_size` and then we can find a split key faster.

Reply:
In our last test, `get_region_approximate_size` may not be as accurate as expected.